validate_markdown_metadata.py

```diff
@@ -1,20 +1,114 @@
-from sys import argv
+import sys
+import yaml
+import re
 
-import frontmatter
+# Helper method that ensures the file has both a starting and ending '---'
+def has_yaml_header(file_content):
+    """ Detects if a string (usually the contents of a .md file) has a YAML header
+    >>> has_yaml_header("---\\nTest: Yes\\n---")
+    True
+    >>> has_yaml_header("Test: No")
+    False
+    >>> has_yaml_header("---\\nProblem: about to be too many dashes\\n----")
+    False
+    """
+    p = re.compile(r"^---$", re.MULTILINE)
+    results = p.findall(file_content)
+    return len(results) > 1
 
-KEYS = ['title', 'datatype', 'sources', 'destinations']
+# This method is extracted from Portmap
+# and it should stay consistent with it if it needs updating
+def extract_yaml_and_body(file_content):
+    """ Extracts a YAML header delimited by lines consisting of '---' from the rest of a markdown document.
+    >>> extract_yaml_and_body("---\\nTest: Data\\nPart: Deux\\n---\\nSeparate this body part\\n")
+    ({'Test': 'Data', 'Part': 'Deux'}, 'Separate this body part\\n')
+    """
+    assert has_yaml_header(file_content) # File does not have a YAML header
+    in_yaml_header = False
+    in_body = False
+    yaml_content = []
+    body_content = []
+    for line in file_content.split("\n"):
+        if not in_yaml_header and not in_body and line == "---":
+            in_yaml_header = True
+        elif in_yaml_header and line == "---":
+            in_yaml_header = False
+            in_body = True
+        elif in_yaml_header:
+            yaml_content.append(line)
+        elif in_body:
+            body_content.append(line)
 
-def article_validator(markdown_file):
-  article = frontmatter.load(markdown_file)
-  keys_indices = {key: index for index, key in enumerate(KEYS)}
-  sorted_keys = sorted(list(article.keys()), key=lambda x: keys_indices.get(x, float('inf')))
+    yaml_content = yaml.safe_load('\n'.join(yaml_content))
+    body = '\n'.join(body_content)
+    return yaml_content, body
 
-  return True if sorted_keys == KEYS else False
+# Checks if a field is of the expected type
+def is_field_valid_type(field_name, field_value, expected_type, file_path):
```

Comment:
Instead of passing "file_path" in here just to be used in the ValueError text, I try to follow the pattern of only passing in the information that's needed to do the functionality.  Then, within validate_frontmatter, the utility method's exception can be caught and the additional information of file_path added at that point.

Reply:
I'll be taking a look at this, but I'll be merging now as we agreed yesterday